onedal/_device_offload.py

```diff
@@ -153,11 +165,36 @@ def _get_host_inputs(*args, **kwargs):
     return q, hostargs, hostkwargs
 
 
-def _extract_usm_iface(*args, **kwargs):
+def _extract_array_attr(*args, **kwargs):
+    """Extracts USM iface, array namespace and Hardware device
+    the array data resides on.
+
+    Returns
+    -------
+    usm_iface : Dict
+        SUA protocol dictionary describing the array.
+    array_api : str
+        The name of the Array API namespace.
+    array_api_device : Array API device object
+        Hardware device the array data resides on.
+
+    """
+
     allargs = (*args, *kwargs.values())
     if len(allargs) == 0:
-        return None
-    return getattr(allargs[0], "__sycl_usm_array_interface__", None)
+        return None, None, None
+    # Getting first argument attr. For all sklearn-like functions
+    # all data provided in the first position. Other data arguments expected
+    # to have the same attributs.
+    firstarg = allargs[0]
+    usm_iface = getattr(firstarg, "__sycl_usm_array_interface__", None)
+    array_api = None
+    array_api_device = None
+    if hasattr(firstarg, "__array_namespace__"):
```

Comment:
Why is a hasattr check necessary?

Reply:
On the line after we are doing call if attr is available

Reply:
https://docs.python.org/3/library/functions.html#getattr

you define a default as None in line 194, meaning it will return None if the attribute isn't available.  Thus the if check on 193 is not necessary.  This is similar to the logic you define in line 190.